src/SignalR/common/Protocols.MessagePack/src/Protocol/MessagePackHubProtocol.cs

```diff
@@ -358,247 +358,236 @@ private static object[] BindArguments(byte[] input, ref int offset, IReadOnlyLis
         /// <inheritdoc />
         public void WriteMessage(HubMessage message, IBufferWriter<byte> output)
         {
-            var writer = MemoryBufferWriter.Get();
+            var writer = new MessagePackWriter(output);
 
-            try
-            {
-                // Write message to a buffer so we can get its length
-                WriteMessageCore(message, writer);
+            // Write message to a buffer so we can get its length
+            WriteMessageCore(message, ref writer);
 
-                // Write length then message to output
-                BinaryMessageFormatter.WriteLengthPrefix(writer.Length, output);
-                writer.CopyTo(output);
-            }
-            finally
-            {
-                MemoryBufferWriter.Return(writer);
-            }
+            // REVIEW : I have no idea if this is needed or not, the previous code was using memryBufferWritter.CopyTo(stream)
```

Comment:
You need the `BinaryMessageFormatter.WriteLengthPrefix` code, it puts the byte count at the beginning of the payload which is needed for the SignalR MessagePack Protocol.

Reply:
You have any suggestion on doing that, since previous code was using `buffer.Length`, and there's not buffer anymore.
Also the `Copy` of the buffer was done after the call to`WriteLengthPrefix`.

Since there is no `buffer` anymore there is no `CopyTo` either, `MessagePack` directly write to the output `IBufferWriter` (zero copy here ?)
Should i take back the buffer here ?

maybe @AArnott have an idea here ?

Reply:
Length prefixing requires serialization in advance. So ya, you need a buffer copy. Using `MemoryBufferWriter` as it did before is probably the simplest way to go.

There are other ways to do this that avoids the copy, but it would be a protocol breaking change for SignalR.

Reply:
thx good to know ;)
Was going to revert with the previous rented buffer as you just said
but as you probably have way more insight here, was waiting for confirmation ^^